test/integration/test_server.cpp

```diff
@@ -1870,6 +1927,130 @@ TEST_CASE("WriteBatchHistory upsert and delete", "[integration][write-batch]") {
+      "transaction_history" + std::to_string(Catch::rngSeed());
+
+  {
+    // Create Tables
+    ::fivetran_sdk::v2::CreateTableRequest request;
+    set_up_plain_write_request(request, MD_TOKEN, TEST_DATABASE_NAME);
+    define_transaction_test_table(request, transaction_table_name);
+
+    ::fivetran_sdk::v2::CreateTableResponse response;
+    auto status = service.CreateTable(nullptr, &request, &response);
+    REQUIRE_NO_FAIL(status);
+
+    ::fivetran_sdk::v2::CreateTableRequest request2;
+    set_up_plain_write_request(request2, MD_TOKEN, TEST_DATABASE_NAME);
+    define_transaction_history_test_table(request2,
+                                          transaction_history_table_name);
+
+    ::fivetran_sdk::v2::CreateTableResponse response2;
+    auto status2 = service.CreateTable(nullptr, &request2, &response2);
+    REQUIRE_NO_FAIL(status2);
+  }
+
+  {
+    // WriteBatch
+    ::fivetran_sdk::v2::WriteBatchRequest request;
+    set_up_plain_write_request(request, MD_TOKEN, TEST_DATABASE_NAME);
+    request.mutable_file_params()->set_unmodified_string(
+        "unmod-NcK9NIjPUutCsz4mjOQQztbnwnE1sY3");
+    request.mutable_file_params()->set_null_string(
+        "null-m8yilkvPsNulehxl2G6pmSQ3G3WWdLP");
+
+    define_transaction_test_table(request, transaction_table_name);
+    request.add_replace_files(TEST_RESOURCES_DIR +
+                              "transaction_input_5_upsert.csv");
+
+    ::fivetran_sdk::v2::WriteBatchResponse response;
+    auto status = service.WriteBatch(nullptr, &request, &response);
+    REQUIRE_NO_FAIL(status);
+  }
+
+  {
+    // WriteHistoryBatch
+    // history write with the earliest file (that does not affect anything
+    // because there is no data), plus upsert file
+    ::fivetran_sdk::v2::WriteHistoryBatchRequest request;
+    set_up_plain_write_request(request, MD_TOKEN, TEST_DATABASE_NAME);
+    request.mutable_file_params()->set_unmodified_string(
+        "unmod-NcK9NIjPUutCsz4mjOQQztbnwnE1sY3");
+    request.mutable_file_params()->set_null_string(
+        "null-m8yilkvPsNulehxl2G6pmSQ3G3WWdLP");
+
+    define_transaction_history_test_table(request,
+                                          transaction_history_table_name);
+    request.add_earliest_start_files(
+        TEST_RESOURCES_DIR + "transaction_history_input_5_earliest.csv");
+    request.add_replace_files(TEST_RESOURCES_DIR +
+                              "transaction_history_input_5_upsert.csv");
+
+    ::fivetran_sdk::v2::WriteBatchResponse response;
+    auto status = service.WriteHistoryBatch(nullptr, &request, &response);
+    REQUIRE_NO_FAIL(status);
+  }
+
+  auto con = get_test_connection(MD_TOKEN);
+  {
+    auto res =
+        con->Query("SELECT id, amount, \"desc\", _fivetran_synced, FROM " +
+                   transaction_table_name + " ORDER BY id");
+    REQUIRE_NO_FAIL(res);
+    REQUIRE(res->RowCount() == 6);
+
+    // All record ids present
+    REQUIRE(res->GetValue(0, 0) == 1);
```

Comment:
(optional) I've started defining helpers for validating expected data like https://github.com/motherduckdb/motherduck-fivetran-connector/blob/38a8cf971392bd84f74e034cc0f7337122ca6988/test/integration/test_server.cpp#L2128-L2134 . It makes it terser and easier to validate multiple rows, too.